src/main/java/com/google/cloud/genomics/dataflow/readers/ReadStreamer.java

```diff
@@ -0,0 +1,144 @@
+ * the License.
+ */
+package com.google.cloud.genomics.dataflow.readers;
+
+import com.google.api.services.genomics.model.ReadGroupSet;
+import com.google.api.services.genomics.model.SearchReadGroupSetsRequest;
+import com.google.cloud.dataflow.sdk.transforms.DoFn;
+import com.google.cloud.dataflow.sdk.transforms.PTransform;
+import com.google.cloud.dataflow.sdk.transforms.ParDo;
+import com.google.cloud.dataflow.sdk.values.PCollection;
+import com.google.cloud.genomics.grpc.Channels;
+import com.google.cloud.genomics.utils.Contig;
+import com.google.cloud.genomics.utils.GenomicsFactory;
+import com.google.cloud.genomics.utils.Paginator;
+import com.google.common.base.Function;
+import com.google.common.collect.FluentIterable;
+import com.google.common.collect.Lists;
+import com.google.genomics.v1.Read;
+import com.google.genomics.v1.StreamReadsRequest;
+import com.google.genomics.v1.StreamReadsResponse;
+import com.google.genomics.v1.StreamingReadServiceGrpc;
+import java.io.IOException;
+import java.security.GeneralSecurityException;
+import java.util.Iterator;
+import java.util.List;
+
+/**
+ * Class with tools for streaming reads using gRPC within dataflow pipelines.
+ */
+public class ReadStreamer {
+
+  /**
+   * Gets ReadGroupSetIds from a given datasetId using the Genomics API.
+   */
+  public static List<String> getReadGroupSetIds(String datasetId, GenomicsFactory.OfflineAuth auth)
+      throws IOException, GeneralSecurityException {
+    List<String> output = Lists.newArrayList();
+    Iterable<ReadGroupSet> rgs = Paginator.ReadGroupSets.create(
+        auth.getGenomics(auth.getDefaultFactory()))
+        .search(new SearchReadGroupSetsRequest().setDatasetIds(Lists.newArrayList(datasetId)),
+            "readGroupSets/id,nextPageToken");
+    for (ReadGroupSet r : rgs) {
+      output.add(r.getId());
+    }
+    if (output.isEmpty()) {
+      throw new IOException("Dataset " + datasetId + " does not contain any ReadGroupSets");
+    }
+    return output;
+  }
+
+  /**
+   * Constructs a StreamReadsRequest for a readGroupSetId, assuming that the user wants all
+   * to include all references.
+   */
+  public static StreamReadsRequest getReadRequests(String readGroupSetId) {
+    return StreamReadsRequest.newBuilder()
+        .setReadGroupSetId(readGroupSetId)
+        .build();
+  }
+
+  /**
+   * Constructs a StreamReadsRequest for a readGroupSetId and a set of user provided references.
+   */
+  public static List<StreamReadsRequest> getReadRequests(final String readGroupSetId,
+      String references) {
+    final Iterable<Contig> contigs = Contig.parseContigsFromCommandLine(references);
+    return FluentIterable.from(contigs)
+        .transformAndConcat(new Function<Contig, Iterable<Contig>>() {
+            @Override
+            public Iterable<Contig> apply(Contig contig) {
+              return contig.getShards(100000000L);
```

Comment:
Ah forgot about that one. I think that's fine to do in a follow-up PR.

On Mon, Jun 22, 2015 at 2:09 PM, Nicole Deflaux <notifications@github.com>
wrote:

> In
> src/main/java/com/google/cloud/genomics/dataflow/readers/ReadStreamer.java
> <https://github.com/googlegenomics/dataflow-java/pull/99#discussion_r32983969>
> :
>
> > +    return StreamReadsRequest.newBuilder()
> > +        .setReadGroupSetId(readGroupSetId)
> > +        .build();
> > +  }
> > +
> > +  /**
> > +   * Constructs a StreamReadsRequest for a readGroupSetId and a set of user provided references.
> > +   */
> > +  public static List<StreamReadsRequest> getReadRequests(final String readGroupSetId,
> > +      String references) {
> > +    final Iterable<Contig> contigs = Contig.parseContigsFromCommandLine(references);
> > +    return FluentIterable.from(contigs)
> > +        .transformAndConcat(new Function<Contig, Iterable<Contig>>() {
> > +            @Override
> > +            public Iterable<Contig> apply(Contig contig) {
> > +              return contig.getShards(100000000L);
>
> @dionloy <https://github.com/dionloy> What about #61
> <https://github.com/googlegenomics/dataflow-java/issues/61>?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/googlegenomics/dataflow-java/pull/99/files#r32983969>.
>